### DISCUSSION TITLE
Phonon relaxation times at certain q point

### DISCUSSION THREAD

**Discussion by biglinn** [SSCHAcode/python-sscha#357] [Q&A]
Hi, @DjordjeDangic
     I find there are outputs `frequency` and `lifetime` in the `Phonon_transport_properties_X` after calculating thermal conductivity, like this:
![image](https://github.com/user-attachments/assets/ef03c7aa-3fdd-4566-ad20-73f4df161d6a)

Since there are thousands of lines that confuses me, could you explain the structure of all lines?
Because I'd like to calculate the phonon relaxation time at certain q point(such as q=0).

By the way, I see that the total phonon relaxation time contain not only anharmonic scattering but also other scattering in _APPENDIX B_ of 
the article _10.1103/PhysRevB.88.045430_.
So dose the `lifetime`  in `Phonon_transport_properties_X` is total phonon relaxation time?

**Comment by DjordjeDangic** [comment]
Each line is one phonon mode. It has thousands of lines because you have thousands of phonon modes in your system (number of q points times the number of bands).

If you save the ThermalConductivity object (with tc.save_pickle() function where tc is your ThermalConductivity object) you can load it and then find all of your calculated properties. This is shown in the tutorial that is on sscha website. For example, if you need lifetimes at Gamma (change temperature accordingly):

```
import numpy as np
import cellconstructor as CC
import cellconstructor.ThermalConductivity

tc = CC.ThermalConductivity.load_thermal_conductivity()

temperature = 400.0
key = format(temperature, '.1f')

for iqpt in range(tc.nkpt):
    if(np.linalg.norm(tc.k_points[iqpt]) == 0.0):
        print('Lifetimes in seconds in k point: ', tc.k_points[iqpt])
        print(tc.lifetimes[key][iqpt])
```

Lifetimes calculated in SSCHA are just due to anharmonicity (just the first term in that reference).

**Reply by biglinn** [reply]
It really is!
I set kgrid mesh=[8,8,8] in the `CC.ThermalConductivity.ThermalConductivity(  `, and it has 12 phonon bands. So there are 6144(512*12) lines.

But what is the structure of the file? 
Is it like that 1-512 lines belong to one band, 513-1024 lines belong to another...
or that 1-12 lines are 12 bands at one point, 13-24 another point...  ?

**Reply by DjordjeDangic** [reply]
The first 12 lines are the first q point (which is usually Gamma). The second 12 lines are from a second q point, etc.

**Comment by biglinn** [comment]
And I'm still not sure about one thing.
I calculate thermal conductivity by following the tutorial which calculates **by potential**(such as ML potential). 
The two important steps of process is:
1, calculate initial dynamical matrices of corresponding supercell by `CC.Phonons.compute_phonons_finite_displacements`
2, sample ensembles and calculate energy and forces by `compute_ensemble( `

Before, I calculated anharmonic dyn following the tutorial _https://sscha.eu/Tutorials/tutorial_02_advanced_submission/_.
In this workflow, **initial dyn come from QE ph.x**. **Energy and forces are calculated by QE pw.x**.

So, I wonder if I can calculate thermal conductivity like this. Because I will use QE to get initial dyn and calculate energy and forces if I research other materials without potential available.
Maybe it's a stupid question, but I think it's best to ask for it.

**Reply by DjordjeDangic** [reply]
You do not have to use machine learning potential to calculate force constants. You can do everything with DFT. However, keep in mind that to converge results you will need to calculate a large number of configurations and this will not be fast with DFT.

**Reply by biglinn** [reply]
I get it. As long as I can get SSCHA dyn to calculate 3rd-order IFCs, it's okay.
Thanks!

**Comment by biglinn** [comment]
I have another question.
The dyn and 3rd-order IFCs are calculated in the same script in the tutorial _https://sscha.eu/Tutorials/tutorial_09_TC/_.
If I have had calculated dyn, could I directly calculate the 3rd-order IFCs? As follows:
```
import sscha, sscha.Ensemble, sscha.SchaMinimizer
import sscha.Utilities
import cellconstructor as CC, cellconstructor.Phonons

dyn = CC.Phonons.Phonons("sscha_dyn", 4)
ensemble = sscha.Ensemble.Ensemble(dyn, 0)

dyn_hessian, d3_tensor = ensemble.get_free_energy_hessian(include_v4 = False, get_full_hessian = True, return_d3 = True)
np.save("d3.npy", d3_tensor)
```


Moreover, before calculating 3rd-order IFCs, there is `update weights` in the script of toturial:
![image](https://github.com/user-attachments/assets/2d0c5713-154b-4794-a22c-a2360b553aad)

What role dose it play? And necessary?

**Reply by biglinn** [reply]
Hi, @DjordjeDangic

After I get SSCHA dyn, I calculate 3rd-order IFCs with the following script:
```
import numpy as np
import sscha, sscha.Ensemble, sscha.SchaMinimizer
import sscha.Utilities
import cellconstructor as CC, cellconstructor.Phonons
from datetime import datetime

dyn = CC.Phonons.Phonons("dyn_start_population1_", 3)
ensemble = sscha.Ensemble.Ensemble(dyn, 0)

time1 = datetime.now()

dyn_hessian, d3_tensor = ensemble.get_free_energy_hessian(include_v4 = False, get_full_hessian = True, return_d3 = True)

time2 = datetime.now()
time_difference = time2 - time1
print(f"hessian needs {time_difference}")

np.save("d3.npy", d3_tensor)
```
But it give erros:
```
Traceback (most recent call last):
  File "/public/home/biglinn/mote2/mote2-tc/hessian.py", line 13, in <module>
    dyn_hessian, d3_tensor = ensemble.get_free_energy_hessian(include_v4 = False, get_full_hessian = True, return_d3 = True)
  File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/sscha/Ensemble.py", line 3644, in get_free_energy_hessian
    self.convert_units(UNITS_HARTREE)
  File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/sscha/Ensemble.py", line 364, in convert_units
    self.forces /= 2 * __A_TO_BOHR__ #Ry/A -> Ha/bohr
TypeError: unsupported operand type(s) for /=: 'list' and 'float'
```

What problem dose this simple script have?

**Reply by DjordjeDangic** [reply]
Yes, update_weights is necessary so that the calculated 3rd-order force constants are the ones that correspond to the final_dyn dynamical matrices. At least, this is what I think, Lorenzo should know better. 

The second issue is probably due to the fact that self.forces is a list and not a numpy array. As to why this happened, I would guess you did not read in the ensemble properly and self.forces is an empty list. Where do you load in the computed ensemble?

**Reply by mesonepigreco** [reply]
@DjordjeDangic is right. update_weights is necessary after performing a minimization (which is always suggested before computing the hessian). As he mentioned, the error you get comes from the fact that the ensemble is empty (you did not load the configurations). I added issue #362 to improve the error message in the next release with a more informative description of what is happening rather than this crash.

**Reply by biglinn** [reply]
I get it. I thought that maybe only minimization needs to load configurations.
And does the temprature for `update_weights ` keep the same with temprature when calculating SSCHA dyn?

would you see if the following script is available?
```
import numpy as np
import sscha, sscha.Ensemble, sscha.SchaMinimizer
import sscha.Utilities
import cellconstructor as CC, cellconstructor.Phonons
from datetime import datetime

dyn = CC.Phonons.Phonons("sscha_dyn_population1_", 3)
ensemble = sscha.Ensemble.Ensemble(dyn, 0)
ensemble.load("/public/home/biglinn/mote2/mote2-dyn/331/data_and_job", population = 1, N=20)
ensemble.update_weights(ensemble.dyn_0, 300)

time1 = datetime.now()

dyn_hessian, d3_tensor = ensemble.get_free_energy_hessian(include_v4 = False, get_full_hessian = True, return_d3 = True)

time2 = datetime.now()
time_difference = time2 - time1
print(f"hessian needs {time_difference}")

np.save("d3.npy", d3_tensor)
```

**Reply by DjordjeDangic** [reply]
Hey biglinn,

I think your calculation will run. However, since I do not know what you want to calculate and in which way the ensemble was obtained (the one at "/public/home/biglinn/mote2/mote2-dyn/331/data_and_job" path) I can not tell you if results are going to make sense and/or they will represent faithfully what you want to calculate. However, I can confidently tell you that N=20 is too small to get anything meaningful.

**Reply by biglinn** [reply]
Hi, @DjordjeDangic

Unfortunately, there is something wrong I don't know what happened.

I list what I do in each step below:
1, I calculate initial dyn by QE ph.x
2, I sample ensembles from the initial dyn:
```
import cellconstructor as CC, cellconstructor.Phonons
import sscha, sscha.Ensemble
import numpy as np

dyn = CC.Phonons.Phonons("harmonic_dyn/MoTe2_harmonic_dyn", 3)
dyn.Symmetrize()
#dyn.ForcePositiveDefinite()
ensemble = sscha.Ensemble.Ensemble(dyn,300)
ensemble.generate(20)
ensemble.save("data_and_job", 1)
```
3, Composing complete input files with `scf_header.in` and configurations to calculate energies and forces by QE
4, Minimization:
```
import sscha, sscha.Ensemble, sscha.SchaMinimizer
import sscha.Utilities
import cellconstructor as CC, cellconstructor.Phonons
from datetime import datetime

POPULATION = 1
dyn = CC.Phonons.Phonons("dyn_mat/dyn_start_population1_", 3)
ensemble = sscha.Ensemble.Ensemble(dyn, 0)
ensemble.load("data_and_job", population = POPULATION, N=20)
time1 = datetime.now()
minim = sscha.SchaMinimizer.SSCHA_Minimizer(ensemble)
minim.set_minimization_step(0.01)
minim.meaningful_factor= 0.001
minim.max_ka = 10000
minim.init()
# Save the minimization details
ioinfo = sscha.Utilities.IOInfo()
ioinfo.SetupSaving("minim_{}".format(POPULATION))
minim.run(custom_function_post = ioinfo.CFP_SaveAll)
minim.finalize()
time2 = datetime.now()
time_difference = time2 - time1
print(f"minimization needs {time_difference}")
minim.dyn.save_qe("dyn_mat/sscha_dyn_population{}_".format(POPULATION))
```
(chang population_id for a few more iterations)
5, Calculate 3rd-order IFCs by SSCHA dyn: (The following script is the one mentioned above)
```
import numpy as np
import sscha, sscha.Ensemble, sscha.SchaMinimizer
import sscha.Utilities
import cellconstructor as CC, cellconstructor.Phonons
from datetime import datetime

dyn = CC.Phonons.Phonons("sscha_dyn_population1_", 3)
ensemble = sscha.Ensemble.Ensemble(dyn, 0)
ensemble.load("/public/home/biglinn/mote2/mote2-dyn/331/data_and_job", population = 1, N=20)
ensemble.update_weights(ensemble.dyn_0, 300)

time1 = datetime.now()

dyn_hessian, d3_tensor = ensemble.get_free_energy_hessian(include_v4 = False, get_full_hessian = True, return_d3 = True)

time2 = datetime.now()
time_difference = time2 - time1
print(f"hessian needs {time_difference}")

np.save("d3.npy", d3_tensor)
```
6, Calculate thermal conductivity:
```
from __future__ import print_function
from __future__ import division
import numpy as np
import cellconstructor as CC
import cellconstructor.Phonons
import cellconstructor.ForceTensor
import cellconstructor.ThermalConductivity
from datetime import datetime

dyn_prefix = 'sscha_dyn_population1_'
nqirr = 3
SSCHA_TO_MS = cellconstructor.ThermalConductivity.SSCHA_TO_MS
RY_TO_THZ = cellconstructor.ThermalConductivity.SSCHA_TO_THZ
dyn = CC.Phonons.Phonons(dyn_prefix, nqirr)
supercell = dyn.GetSupercell()

fc3 = CC.ForceTensor.Tensor3(dyn.structure,
dyn.structure.generate_supercell(supercell), supercell)
d3 = np.load("d3.npy")
fc3.SetupFromTensor(d3)
fc3 = CC.ThermalConductivity.centering_fc3(fc3)

mesh = [8,8,8]
smear = 0.03/RY_TO_THZ
tc = CC.ThermalConductivity.ThermalConductivity(dyn, fc3,
kpoint_grid = mesh, scattering_grid = mesh, smearing_scale = None,
smearing_type = 'constant', cp_mode = 'quantum', off_diag = False)
temperatures = np.arange(300, 301, 100, dtype=float)

time1 = datetime.now()
tc.setup_harmonic_properties(smear)
tc.write_harmonic_properties_to_file()
tc.calculate_kappa(mode = 'SRTA', temperatures = temperatures,
write_lifetimes = True, gauss_smearing = True, offdiag_mode = 'wigner',
kappa_filename = 'Thermal_conductivity_SRTA', lf_method = 'fortran-LA')
time2 = datetime.now()
time_difference = time2 - time1
print(f"Calculated SSCHA kappa with SRTA needs {time_difference}")
```

**Reply by biglinn** [reply]
I thought it would work. But the step of calculating thermal conductivity crashes with following erros:

![image](https://github.com/user-attachments/assets/abc64130-3c4e-428d-b411-dc468c1fc8e3)

You mentioned this error in https://github.com/orgs/SSCHAcode/discussions/354#discussioncomment-10797953. 
But I plot the phonon dispersion which don't have imaginary frequencies around Gamma point:

![population1](https://github.com/user-attachments/assets/7dc56d12-4ab0-4a86-8cec-a53cd0782240)

I also do several SSCHA iterations and use dyn of different iteration to calculate thermal conductivity. They give the same erros and don't have imaginary frequencies too.
I'm confused of what problems happened.

**Reply by biglinn** [reply]
I find one thing. I plot the dispersion of `hessian dyn` output together with d3.npy.
Although the SSCHA dyn used to calculate 3rd-order IFCs don't have imaginary frequencies, **the hessian dyn does have**:
![hessian_by_load](https://github.com/user-attachments/assets/8560c7f7-e470-452b-8882-cdf9d0b456fc)

I think it may be too less configurations loaded.
So I use `ensemble.generate(5000, sobol = True)` instead of `ensemble.load("/public/home/biglinn/mote2/mote2-dyn/331/data_and_job", population = 1, N=20)` when calculating 3rd-order IFCs.
In this situation, hessian dyn don't have imaginary frequencies:
![hessian](https://github.com/user-attachments/assets/4320f969-9856-4389-9229-6742468dfaca)

But **there is still erro** `w_q = np.sqrt(w2_q)` when calculating thermal conductivity.

**Reply by DjordjeDangic** [reply]
So the error is there because you have an imaginary frequency in whatever grid you are using. But you know this already.

I can imagine two scenarios. In one you are not using the same dynamical matrices to plot phonon band structure and to calculate thermal conductivity. In the other, the imaginary frequency is not on the M-G-K-M path, but somewhere else in the Brillouin zone. Plot the phonon density of states and see if you have DOS weight for the negative frequencies.

**Reply by biglinn** [reply]
I think the first scenario you said should not happen since I have checked this several times.
About the second scenario, why I plot just on M-G-K-M path is that the **MoTe2 is 2D material** and **M-G-K-M path is in a plane** of Brillouin zone. And I also plot on path G-M-K-G-A-L-H-A(this path is recommended by Materials Cloud), as follows:
![population1-long_path](https://github.com/user-attachments/assets/d2bea1f2-ee06-43e6-bdef-ef1bf1d9015f)

So, I don't know where imaginary frequency is or if it exists.

Moveover, I have calculated initial harmonic dyn by QE with two q-grid (**331 and 441**). So corresponding size of supercell is 331 and 441 respectively. Is it possible that q-grid of hamonic dyn is so small that **harmonic dyn don't contain useful information**?

**Reply by biglinn** [reply]
And I try to plot dispersion by QE `q2r.x and matdyn.x`.
![contrast](https://github.com/user-attachments/assets/1c2fe3c7-a9be-45bb-8bc9-4dbf9d2fcb7a)

As you can see, the **second** phonon mode from the top is different **around Gamma point** in two ways.
But this mode is unlikely to have imaginary frequency, isn't it?  It's crazy.

**Reply by mesonepigreco** [reply]
The Gamma point mismatch seems likely to be a problem with LO-TO splitting, effective charges and dielectric tensors

**Reply by biglinn** [reply]
What could be the problem? I check that harmonic dyn by QE have effective charges and dielectric tensors.

And would the acoustic branch mismatch lead to problem of calculating thermal conductivity ?

**Reply by DjordjeDangic** [reply]
I have never worked with 2D materials, so I am not sure whether there are some issues with them. Also, what is A point? Why do you have three modes going to zero there? If you have 2D material why are you using an 8x8x8 q-grid?

**Reply by biglinn** [reply]
A point is (0, 0, 0.5). The 0.5 is in the direction perpendicular to 2D. So A point is  very close to Gamma point.

I think you have found the key to my problem! I didn't realize it. 8x8x8 q-grid may be problematic!

**Reply by biglinn** [reply]
Hi, @DjordjeDangic

I have set the q-grid to `X by X by 1`, namely that all q-points are in the same plane of Brillouin zone. It dose work and calculation of thermal conductivity cat run.

But there is also a problem. 
I have test q-grid 441, 881, 12x12x1, etc. The output shows that **not all q-point** can be calculated, as follows:
![image](https://github.com/user-attachments/assets/dc7dd8ee-14d6-4559-b097-8b6301c3c304)

According to the erro, it seems that **the dyn does have imaginary frequency in some area of Brillouin zone**. But why doesn't phonon dispersion spectrum show it?

**Reply by biglinn** [reply]
![population4](https://github.com/user-attachments/assets/5e4fd73b-46c5-45f6-b8e0-8a16d8ad4567)

With q-grid X by X by 1 (such as 881, 12x12x1), all q points are **in the plane of Gamma point**.
And the M-G-K-M path is **in the same plane**, because of M (0.5, 0.5, 0) and K (0.3333, 0.3333, 0).

Why doesn't the imaginary frequency show at M-G-K-M path of dispersion spectrum?

**Reply by biglinn** [reply]
I have checked the dynamical matrix files (**the first file, corresponding to Gamma point**).
The frequencies of first phonon mode are as follows:

Harmonic dyn by QE:
`freq (    1) =      -0.162911 [THz] =      -5.434134 [cm-1]`

SSCHA dyn after **one** iteration, `sscha_dyn_population1_1`:
`freq (    1) =    -0.00000008 [THz] =    -0.00000277 [cm-1]`

SSCHA dyn after **two** iterations, `sscha_dyn_population2_1`:
`freq (    1) =    -0.00000003 [THz] =    -0.00000114 [cm-1]`

SSCHA dyn after **three** iterations, `sscha_dyn_population3_1`:
`freq (    1) =    -0.00000011 [THz] =    -0.00000358 [cm-1]`

SSCHA dyn after **four** iterations, `sscha_dyn_population4_1`:
`freq (    1) =     0.00000004 [THz] =     0.00000146 [cm-1]`

The imaginary frequency dose exist, just not showed by phonon dispersion.

Although the first mode at Gamma doesn't have imaginary frequency after four iterations, I guess that imaginary frequency may not be eliminated at other area of Brillouin zone. So, perhaps I need more iterations and configurations to get converged SSCHA dyn.

**Reply by DjordjeDangic** [reply]
As I have told you in one of the previous answers, the best way to check for the imaginary frequencies is to plot the phonon density of states calculated with the same grid you are attempting the thermal conductivity calculation.

Also make sure that the scattering grid is also with only one point in the third direction, ie. 8x8x1.

**Reply by biglinn** [reply]
Oh, sorry, I missed that answer. Dose SSCHA provide codes to plot phonon density of states?

**Reply by biglinn** [reply]
I use QE to plot phonon pdos.
I plot pdos of SSCHA dyn with one iteration and with four iterations, as follows:
![phonon_pdos](https://github.com/user-attachments/assets/a1df94ee-0aee-49ed-915b-082ad14e86b1)


It doesn't have DOS weight for the negative frequencies. So strange.
